### DISCUSSION TITLE
Setting compilerArgs when generating tasks.json file for compiling

### DISCUSSION THREAD

**Discussion by djairao** [microsoft/vscode-cpptools#7851] [Q&A] [answered]
Hello,

I have a scenario where I will have many students creating several C++ projects in VS-Code. To streamline the experience, I am trying to set default flags for `g++` so that they are automatically included when the `tasks.json` file is generated. I tried setting the defaults in several spots and it did not seem to work. After mucking with settings for a few hours, I just decided to modify the following line of javascript with the additional flags in the file `code-server/extensions/ms-vscode.cpptools-1.5.1/dist/main.js`

`
             let args = isCl ? ['/Zi', '/EHsc', '/nologo', '/Fe:', filePath + '.exe', '${file}'] : ['-g', **'-Wall', '-std=c++17', '-fsanitize=address'**, '${file}', '-o', filePath + (isWindows ? '.exe' : '')];

                if (compilerArgs && compilerArgs.length > 0) {
                    args = args.concat(compilerArgs);
                }
`
 

That seems to do the trick. Now, I noticed that the configuration variable `compilerArgs` is being used in the above source. However, I am unsure where to set a value for it. Setting it in the `User/settings.json` file as shown below does not help:

`
{
    "window.menuBarVisibility": "classic",
    "editor.renderControlCharacters": true,
    "editor.renderWhitespace": "all",
    "editor.rulers": [
        80
    ],
    "C_Cpp.clang_format_fallbackStyle": "{ BasedOnStyle: Google, IndentWidth: 4}",
    "C_Cpp.autocompleteAddParentheses": true,
    "C_Cpp.default.cppStandard": "c++17",
    "C_Cpp.default.includePath": ["${myDefaultIncludePath}", "/opt/mvapich2/gnu/8.4/2.3.3/include"],
    "**C_Cpp.default.compilerArgs": [ "-std=c++17", "-DONE", "-DTwo"]**,
 
`

I will be thankful for any tips or ideas on how the `compilerArgs` parameter in the above code fragment should be set?

**Comment by sean-mcmanus** [comment, marked as answer]
You shouldn't probably be modifiying the javascript. You could git clone our repo and put a breakpoint at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/cppBuildTaskProvider.ts#L169 to see what is going on -- looks like it's just "undefined". In general, the cppbuild feature (which controls building) wasn't designed to inherit from c_cpp_properties.json or the C_Cpp.default settings (which just controls IntelliSense)....maybe we could accept change to the design, not sure.

If you want to give your students a really streamlined experience, I would recommend setting the configurationProvider and using the CMake Tools or Makefile Tools extensions or compileCommands, i.e. have the build system configure IntelliSense instead of the other way around.

**Comment by djairao** [comment]
Thank you for your quick response. 

Maybe having a configuration variable to set up compiler args would be a good feature/idea. At least by default, I think it would be better to keep things consistent by using the same settings as IntelliSense, if the compiler args are undefined. In addition, it would be nice to have a comment in the generated `tasks.json` file that indicates where the properties were obtained from (for non-default cases) as in:

```json
    "tasks": [
        {
            "type": "cppbuild",
            "label": "C/C++: g++ build active file",
            "command": "/apps/gnu/8.4.0/bin/g++",
             // Overrides from settings.json
            "args": [  "-g",  "-Wall",   "-std=c++17",  "-fsanitize=address",
                "${file}",
                "-o",
                "${fileDirname}/${fileBasenameNoExtension}"
            ],
```

**Reply by sean-mcmanus** [reply]
Did you want to file a feature request? https://github.com/microsoft/vscode-cpptools/issues/new?assignees=&labels=&template=feature-request.md&title=

**Reply by djairao** [reply]
Thank you for the link. I created a new issue -- https://github.com/microsoft/vscode-cpptools/issues/7860

**Comment by djairao** [comment]
I am looking into `vscode-cmake-tools` and it looks like I have to use the command palette to access cmake-tool's features; there aren't any main menu options for this (which would make life a lot easier for novice users). Is this correct (i.e., there are no main menu options) or am I missing something? (I know this is not the `vscode-cmake-tool` discussions, but thought I would start here, in case someone already knows the answer)

**Reply by sean-mcmanus** [reply]
VS Code's design doesn't allow extensions to add to the main menu options -- everything goes through the command palette or potentially other UI, such as the task/status bar UI or the right-click menu or code action menus etc.. VS Code expects novice users to learn to use the Command Palette.

**Reply by djairao** [reply]
To me, the command-pallet seems similar to the commands from `emacs` :smiley_cat:
